dd-sdk-android/src/main/kotlin/com/datadog/android/log/internal/logger/LogcatLogHandler.kt

```diff
@@ -22,21 +25,61 @@ internal class LogcatLogHandler(
         attributes: Map<String, Any?>,
         tags: Set<String>
     ) {
+        val tag = resolveTag()
         if (Build.MODEL == null) {
-            println("${levelPrefixes[level]}/$serviceName: $message")
+            println("${levelPrefixes[level]}/$tag: $message")
             throwable?.printStackTrace()
         } else {
-            Log.println(level, serviceName, message)
+            Log.println(level, tag, message)
             if (throwable != null) {
                 Log.println(
                     level,
-                    serviceName,
+                    tag,
                     Log.getStackTraceString(throwable)
                 )
             }
         }
     }
 
+    private fun resolveTag(): String {
+        return if (Datadog.isDebug) {
+            return resolveTagFromCallerClassName()
+        } else {
+            serviceName
+        }
+    }
+
+    private fun resolveTagFromCallerClassName(): String {
+        val stackTrace = Throwable().stackTrace
+        if (stackTrace.size <= callerNameStackIndex) {
+            return serviceName
+        }
+
+        // remove the Anonymous class name
+        val className = stackTrace[callerNameStackIndex].className
+        var tag = stripAnonymousPart(className)
+        tag = tag.substring(tag.lastIndexOf('.') + 1)
+
+        return sanitizeTag(tag)
+    }
+
+    private fun sanitizeTag(tag: String): String {
+        return if (tag.length < MAX_TAG_LENGTH || Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            tag
+        } else {
+            tag.substring(0, MAX_TAG_LENGTH)
+        }
+    }
+
+    private fun stripAnonymousPart(className: String): String {
+        val matcher = ANONYMOUS_CLASS.matcher(className)
+        return if (matcher.find()) {
```

Comment:
You don't need the if block . `matcher.replaceAll()` will reset the matcher state and parse again the input string, meaning you're searching for the pattern twice (once in find and another one in replaceAll). 
If nothing is found, it won't replace anything anyway.